examples/cfn-custom-resource/ami-lookup.js

```diff
@@ -0,0 +1,103 @@
+/**
+* A sample Lambda function that looks up the latest AMI ID for a given region and architecture.
+**/
+
+// Map instance architectures to an AMI name pattern
+var archToAMINamePattern = {
+    "PV64": "amzn-ami-pv*x86_64-ebs",
+    "HVM64": "al2023-ami-2023.*-kernel-*-x86_64",
+    "HVMG2": "amzn-ami-graphics-hvm*x86_64-ebs*"
+};
+const { EC2Client, DescribeImagesCommand } = require("@aws-sdk/client-ec2");
+
+exports.handler = async function(event, context) {
+    const redactedEvent = { ...event, ResponseURL: "REDACTED" };
+    console.log("REQUEST RECEIVED:\n" + JSON.stringify(redactedEvent));
+    
+    // For Delete requests, immediately send a SUCCESS response.
+    if (event.RequestType == "Delete") {
+        await sendResponse(event, context, "SUCCESS");
+        return;
+    }
+ 
+    var responseStatus = "FAILED";
+    var responseData = {};
+ 
+    const ec2Client = new EC2Client({ region: event.ResourceProperties.Region });
+    const describeImagesParams = {
+        Filters: [{ Name: "name", Values: [archToAMINamePattern[event.ResourceProperties.Architecture]]}],
+        Owners: [event.ResourceProperties.Architecture == "HVMG2" ? "679593333241" : "amazon"]
+    };
+ 
+    try {
+        const describeImagesResult = await ec2Client.send(new DescribeImagesCommand(describeImagesParams));
+        var images = describeImagesResult.Images;
+        // Sort images by name in descending order. The names contain the AMI version, formatted as YYYY.MM.Ver.
+        images.sort((x, y) => y.Name.localeCompare(x.Name));
+        for (var j = 0; j < images.length; j++) {
+            if (isBeta(images[j].Name)) continue;
+            responseStatus = "SUCCESS";
+            responseData["Id"] = images[j].ImageId;
+            break;
+        }
+    } catch (err) {
+        responseData = { Error: "DescribeImages call failed" };
+        console.log(responseData.Error + ":\n", err);
+    }
+    
+    await sendResponse(event, context, responseStatus, responseData);
+};
+
+// Check if the image is a beta or rc image. The Lambda function won't return any of those images.
+function isBeta(imageName) {
+    return imageName.toLowerCase().indexOf("beta") > -1 || imageName.toLowerCase().indexOf(".rc") > -1;
+}
+
+// Send response to the pre-signed S3 URL 
+async function sendResponse(event, context, responseStatus, responseData) {
```

Comment:
I am curious here, when people write lambdas to power custom resources in CloudFormation, do they need to take care of this functionality in user code or is there some library or framework that users can call in CF?

Reply:
This needs to be part of the user code in the lambda function.

When using Inline Code in CloudFormation you can use the `cfn-response` module if it's a node Lambda. That's automatically included in the Lambda Zip when using inline code.
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-lambda-function-code-cfnresponsemodule.html#cfn-lambda-function-code-cfnresponsemodule-source

Reply:
You could use it in the example too presumably? https://www.npmjs.com/package/cfn-response seems available. No big deal though

Reply:
That's not the official version. Somebody just extracted it and published it on Lambda. I wouldn't wanna rely on this in-official version for the example